router/h2/src/main/scala/io/buoyant/router/h2/StreamStatsFilter.scala

```diff
@@ -1,83 +1,130 @@
       durationMs.add(startT().inMillis)
       result match {
         case Return(_) => successes.incr()
         case Throw(_) => failures.incr()
       }
     }
+
+    @inline def apply(startT: Stopwatch.Elapsed)(result: Try[_]): Unit = onEnd(startT)(result)
+
+  }
+
+  trait FrameStats extends StreamStats {
+    private[this] val frameSize = stats.stat("data_frame", "bytes")
+    private[this] val frameCount = stats.stat("data_frame", "count")
+
+    def onFrame(startT: Stopwatch.Elapsed)(stream0: Stream): Stream = {
+      var streamFrameCount = 0
+      val stream1 = stream0.onFrame {
+        case Return(frame: Frame.Data) =>
+          frameSize.add(frame.buf.length)
+          streamFrameCount += 1
+        case _ =>
+      }
+      val _ = stream1.onEnd.respond { result =>
+        frameCount.add(streamFrameCount)
+        this.onEnd(startT)(result)
+      }
+      stream1
+    }
+
+    @inline def apply(startT: Stopwatch.Elapsed)(stream: Stream): Stream = onFrame(startT)(stream)
+  }
+
+  object ClassifierStats {
+    // overall successes stat from response classifier
+    private[this] val successes = statsReceiver.counter("successes")
+    // overall failures stat from response classifier
+    private[this] val failures = statsReceiver.counter("failures")
+
+    def apply(reqrep: H2ReqRep): Unit = classifier(reqrep) match {
+      case Successful(_) => successes.incr()
+      case _ => failures.incr()
+    }
   }
 
   //   total number of requests received
   private[this] val reqCount = statsReceiver.counter("requests")
 
   private[this] val reqStreamStats =
-    new StreamStats(statsReceiver.scope("request", "stream"))
+    new StreamStats(statsReceiver.scope("request", "stream")) with FrameStats
   private[this] val rspStreamStats =
-    new StreamStats(statsReceiver.scope("response", "stream"))
+    new StreamStats(statsReceiver.scope("response", "stream")) with FrameStats
   private[this] val totalStreamStats =
-    new StreamStats(statsReceiver.scope("stream"),
-                    durationName = Some("total_latency"))
-  private[this] val rspFutureStats =
-    new StreamStats(statsReceiver,
-                    durationName = Some("request_latency"),
-                    successName = Some(""))
-
-  override def apply(req: Request, service: Service[Request, Response]): Future[Response] = {
+    new StreamStats(
+      statsReceiver.scope("stream"),
+      durationName = Some("total_latency")
+    )
+
+  private[this] val reqLatency = statsReceiver.stat("request_latency_ms")
+
+  override def apply(req0: Request, service: Service[Request, Response]): Future[Response] = {
     reqCount.incr()
     val reqT = Stopwatch.start()
+    val req1 = Request(req0.headers, reqStreamStats(reqT)(req0.stream))
 
-    req.stream.onEnd.respond(reqStreamStats(reqT))
-
-    service(req)
+    service(req1)
+      .transform {
+        case Return(rsp0) =>
+          val rspT = Stopwatch.start()
+          val stream = rsp0.stream.onFrame {
```

Comment:
I think we also need to handle the case where the stream isEmpty separately.  I think it should add(0) to the frameCount stat and should call the classifier with None as the final frame.